Add subcategories definitions

jsoriano opened the pull request:
First stab to subcategories support.

By now it only models the categories with the subcategories, loading them from a yaml file. We can consider moving this yaml file to the spec so it can be also consumed from other places.

This PR doesn't include the changes in the API, this will come in a follow-up.

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-914/detail/PR-914/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-914/detail/PR-914/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-914/detail/PR-914/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-914/detail/PR-914/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://package-registry_914.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2022-11-24T09:33:58.320Z&rangeTo=2022-11-24T09:53:58.320Z&transactionName=BUILD+Ingest-manager%2Fpackage-registry%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=e8e741c313485c64313da3ce335d20cd&transactionId=828790af67881eac)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-11-24T09:43:58.320+0000


* Duration: 3 min 6 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-914/detail/PR-914/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Checks formatting / linting`
<ul>
<li>Took 0 min 22 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/package-registry/pipelines/PR-914/runs/1/steps/135/log/?start=0">here</a></li>
<li>Description: <code>mage -debug check</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

mrodm commented on categories/categories.go:
```diff
@@ -0,0 +1,72 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode categories: %w", err)
+	}
+	// TODO: Check for duplicated categories.
+	return categoriesFile.Categories, nil
```
I was thinking just about those possible duplicated categories. Could a category be defined as a subcategory of more than category?

jsoriano replied:
This won't be possible with the current approach. We will need to use conventions (like `network_security` vs. `network_monitoring`) if we want something to be a subcategory in two different areas.

jlind23 commented on categories/categories.yml:
```diff
@@ -0,0 +1,143 @@
+    subcategories:
+      app_search:
+        title: "Application Search"
+      connector: # TODO: Do we want three different connector categories?
```
@mukeshelastic @nimarezainia would you be able to weight in here? Do we want three categories?

mukeshelastic replied:
yes, let's limit the subcategories to only one level depth.. category -> subcategory. And author can specify only one subcategory

jsoriano replied:
> And author can specify only one subcategory

Does it mean that a package can have only a subcategory? We don't have this limitation with categories, and this may complicate things.

jlind23 reviewed (commented):
Should we state somewhere in the code that categories shouldn't have more than 1 nested level for now?

jsoriano commented:
> Should we state somewhere in the code that categories shouldn't have more than 1 nested level for now?

Current code doesn't support nesting categories. I have added also a comment to the header of the yaml file.

jsoriano commented on categories/categories.go:
```diff
@@ -0,0 +1,72 @@
+package categories
+
+import (
+	_ "embed"
```
Fixed. Thanks.

jsoriano commented:
Updated list of categories adding the ones I missed.

There are still some open questions, they are as TODOs in the yml, but grouping them:
1. New subcategories that were main categories before. Options:
    1. Remove main category, leave the subcategory. I'd prefer this option. To avoid breaking changes this would require the API design with the [`subcategory_of`](https://github.com/elastic/package-spec/issues/451#issuecomment-1324172123) field, what is also fine for me.
    2. Keep main category, add a new subcategory with a different name.
2. New categories that overlap with old categories (such as `datastore` vs. `database`). Options:
    1. Keep old categories, dismiss new ones.
    2. Keep old categories, but change the title to the new one.
    3. Keep both categories. I'd avoid this option as can be confusing.
3. Duplicated subcategories (such as `observability/monitoring` vs. `infrastructure/monitoring`. Options:
    1. Keep subcategories with different suffixes. I have already added suffixes to some categories.
    2. Use a single category. We would need to decide which one to keep.
    3. Decide case by case between the previous options.
    4. Allow subcategories to be in multiple main categories. This can be ambiguous as described in https://github.com/elastic/package-registry/pull/914#discussion_r1031388151
4. Too specific product categories. Such as the one for WebSphere, but Kubernetes could be another one. Options:
    1. Keep them. This is probably fine, but where to draw the line before adding more.
    2. Keep them, but only as subcategories.
    3. Decide case by case.

@jlind23 @mukeshelastic @dborodyansky please take a look to these open questions. I can also take the decision, but I may be overlooking something.

amitkanfer commented:
Maybe we should ask the relevant peers from the solution teams to review and signoff?
Also - i guess it's not a problem to change it later on...? @jsoriano

jlind23 commented:
@jsoriano my answers below

> 1. New subcategories that were main categories before. Options:
    1. Remove main category, leave the subcategory. I'd prefer this option. To avoid breaking changes this would require the API design with the [`subcategory_of`](https://github.com/elastic/package-spec/issues/451#issuecomment-1324172123) field, what is also fine for me.

This one will provide more clarity and avoid duplication problems in the near future.

> 2. New categories that overlap with old categories (such as `datastore` vs. `database`). Options:
    1. Keep old categories, dismiss new ones.
    2. Keep old categories, but change the title to the new one.
    3. Keep both categories. I'd avoid this option as can be confusing.

Looking at the work done by @dborodyansky I would rather remove the old categories and display the ones. @kpollich could this be a breaking change on the UI end? Don't we have some hard coded featured integration that relies on some categories?

> 3. Duplicated subcategories (such as `observability/monitoring` vs. `infrastructure/monitoring`. Options:
    1. Keep subcategories with different suffixes. I have already added suffixes to some categories.
  
Keeping subcategories with different suffixes is the best path forward, we might have observability users eager to find monitoring related integrations under the observability umbrella and who won't take a look under the infrastructure umbrella for example.

> 4. Too specific product categories. Such as the one for WebSphere, but Kubernetes could be another one. Options:
    1. Keep them. This is probably fine, but where to draw the line before adding more.
   
We should keep them for now. The line about adding more categories was already explained by Akshay in this document:
https://docs.google.com/document/d/1pMsELVC4XbGXeFsR2gvELk5HLrLaNCujztogRzKjbOU/edit

kpollich commented:
> Looking at the work done by @dborodyansky I would rather remove the old categories and display the ones. @kpollich could this be a breaking change on the UI end? Don't we have some hard coded featured integration that relies on some categories?

There is some hardcoding related to categories in the custom integrations plugin here: https://github.com/elastic/kibana/blob/2c774f536e2cd0b01b8829d842684e67df2c3d57/src/plugins/custom_integrations/common/index.ts

We'll need to align this file to the new categories/subcategories concept in general, but if there are any renames we should tackle those first to avoid breakage.

jsoriano commented:
Thanks @jlind23 for your clarifications, a couple of follow ups:

> > 1. New subcategories that were main categories before. Options:
> >    1. Remove main category, leave the subcategory. I'd prefer this option. To avoid breaking changes this would require the API design with the [`subcategory_of`](https://github.com/elastic/package-spec/issues/451#issuecomment-1324172123) field, what is also fine for me.
> 
> This one will provide more clarity and avoid duplication problems in the near future.
> 

:+1: 

> > 2. New categories that overlap with old categories (such as `datastore` vs. `database`). Options:
> >    1. Keep old categories, dismiss new ones.
> >    2. Keep old categories, but change the title to the new one.
> >    3. Keep both categories. I'd avoid this option as can be confusing.
> 
> Looking at the work done by @dborodyansky I would rather remove the old categories and display the ones. @kpollich could this be a breaking change on the UI end? Don't we have some hard coded featured integration that relies on some categories?
>

Removing categories is not an option, at least without breaking changes. The closer option would be option two, keeping old categories, but changing the titles.
There could be a similar fourth option, that would be to keep both old and new categories, but both with the new titles.
 
> > 3. Duplicated subcategories (such as `observability/monitoring` vs. `infrastructure/monitoring`. Options:
> >    1. Keep subcategories with different suffixes. I have already added suffixes to some categories.
> 
> Keeping subcategories with different suffixes is the best path forward, we might have observability users eager to find monitoring related integrations under the observability umbrella and who won't take a look under the infrastructure umbrella for example.
> 

:+1: 

Regarding the current main `monitoring` category, should we move it under `observability`, so we have:
* `monitoring` as subcategory of `observability`.
* `monitoring_infrastructure` as subcategory of `infrastructure`.

Or keep the three categories:
* `monitoring` main category as the current one.
* `monitoring_observability` as subcategory of `observability`.
* `monitoring_infrastructure` as subcategory of `infrastructure`.

> > 4. Too specific product categories. Such as the one for WebSphere, but Kubernetes could be another one. Options:
> >    1. Keep them. This is probably fine, but where to draw the line before adding more.
> 
> We should keep them for now. The line about adding more categories was already explained by Akshay in this document: https://docs.google.com/document/d/1pMsELVC4XbGXeFsR2gvELk5HLrLaNCujztogRzKjbOU/edit

:+1:

jsoriano commented:
TODOs resolved, and API changed, please review.

jsoriano commented on CHANGELOG.md:
```diff
@@ -8,11 +8,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Breaking changes
 
+* Updated titles of some categories. [#914](https://github.com/elastic/package-registry/pull/914)
```
This shouldn't be a breaking change, but not sure if clients are using the titles for something apart of showing them. 

@kpollich do you know if the integrations UI uses the titles for something apart of to show them?

kpollich replied:
I don't think so - seems like the custom integrations plugin relies on the category ID's for its logic, and Integrations UI just uses the title for disply purposes https://github.com/elastic/kibana/blob/2c774f536e2cd0b01b8829d842684e67df2c3d57/src/plugins/custom_integrations/common/index.ts

jsoriano commented on testdata/generated/storage-indexer/categories.json:
```diff
@@ -81,7 +84,8 @@
+    "title": "Web Server",
+    "count": 14,
+    "subcategory_of": "infrastructure"
   }
```
These will be the changes in the API:
* Titles updated for some categories.
* `subcategory_of` added in subcategories.

jsoriano commented on testdata/generated/categories-experimental.json:
```diff
@@ -31,22 +31,25 @@
-    "count": 3
+    "title": "Web Server",
+    "count": 3,
+    "subcategory_of": "infrastructure"
```
@kpollich with this API response, Fleet is not going to be able to get the title of the parent category, should I change this to something like this?
```suggestion
    "parent_id": "infrastructure",
    "parent_name": "Infrastructure"
```

An alternative could be to include the parent category in the response, but this could affect current versions of kibana that would receive categories with count 0.

This also makes me wonder if we want to modify the search APIs, to include packages of subcategories, when searching for a main category. That would for example be to include packages of the `web` subcategory, when looking for the `integration` category.

kpollich replied:
> with this API response, Fleet is not going to be able to get the title of the parent category, should I change this to something like this?

It seems like the issue is that integrations can be placed in a subcategory even if there are no integrations placed in the parent category, resulting in a case where the parent category is not included in the response. 

So we have a state like this

![image](https://user-images.githubusercontent.com/6766512/204870731-f79084dc-e2fa-4681-92f5-7712211c37c5.png)

I think returning the `parent_name` would be most helpful here, as it allows the UI to render the `Infrastructure` category in this case even if it does not directly contain any integrations.

jsoriano replied:
PR updated to include the `parent_name` in the API response, @kpollich please take another look.
